### PR TITLE
Add option to view diff between revisions

### PR DESCRIPTION
**Pull request by xphoniex**
On GitHub, you see diff (or new commits pushed) through notification link, and then there's the compare button:
Screenshot from 2023-05-16 17-11-36

Here's a button `diff` added to each revision (except first one, obviously):

Screenshot from 2023-05-17 06-35-06

- We can change the name from `diff` to `changeset` or `changes` too.
- We can also use an icon, or an icon of text (e.g. some edgy mono font, like the one used in radicle logo)

Upon clicking on a diff, you're taken to `../<rid>/patches/<cob>?diff=h1..h2`, e.g.:
```
https://app.local/seeds/radicle.local/
  rad:z3gqcJUoA1n9HaHKufZs5FCSGazv5/
  patches/
  84746b79c89de575d0023e64bccca93774eaab75?diff=969ba8633207bf92ac60952ace310f672d774444..423809c05c916c3b2bf2a03bd05889eb4792039f
```

Screenshot from 2023-05-17 06-42-33

- Notice "Activity" tab is still active here
- Before the diff view, we are seeing the patch description, we can either keep that or show the Revision description there, or maybe the Revision and its previous one:

Screenshot from 2023-05-17 06-45-15

So instead of:
```
Add alias to cli tooling ...
```
it can be:
```
- use `profile.home.node().join(TRACKING_DB_FILE)`
```
or:
```
| - use `profile.home.node().join(TRACKING_DB_FILE)`  | vs | Use [ .. ] instead of concrete times in tests |
```

### CODE REVIEW

**Comment by rudolfs** [comment]
> We can change the name from diff to changeset or changes too.

My preference would be if we hide it in a context dropdown.
This way we have a place to potentially add more actions in the future without cluttering the UI too much.
Screenshot 2023-05-17 at 10 15 49

```diff
diff --git a/src/components/Icon.svelte b/src/components/Icon.svelte
index fd49098e..e2278103 100644
--- a/src/components/Icon.svelte
+++ b/src/components/Icon.svelte
@@ -13,6 +13,7 @@
     | "clipboard"
     | "clipboard-small"
     | "ellipsis"
+    | "ellipsis-small"
     | "exclamation"
     | "exclamation-circle"
     | "fork"
@@ -104,6 +105,19 @@
       10.6716 11.1716 10 12 10C12.8284 10 13.5 10.6716 13.5 11.5ZM20 11.5C20
       12.3284 19.3284 13 18.5 13C17.6716 13 17 12.3284 17 11.5C17 10.6716
       17.6716 10 18.5 10C19.3284 10 20 10.6716 20 11.5Z" />
+  {:else if name === "ellipsis-small"}
+    <path
+      fill-rule="evenodd"
+      clip-rule="evenodd"
+      d="M7.66663 12.6667C8.21891 12.6667 8.66663 12.219 8.66663
+      11.6667C8.66663 11.1144 8.21891 10.6667 7.66663 10.6667C7.11434 10.6667
+      6.66663 11.1144 6.66663 11.6667C6.66663 12.219 7.11434 12.6667 7.66663
+      12.6667ZM13 11.6667C13 12.219 12.5522 12.6667 12 12.6667C11.4477 12.6667
+      11 12.219 11 11.6667C11 11.1144 11.4477 10.6667 12 10.6667C12.5522
+      10.6667 13 11.1144 13 11.6667ZM17.3333 11.6667C17.3333 12.219 16.8856
+      12.6667 16.3333 12.6667C15.781 12.6667 15.3333 12.219 15.3333
+      11.6667C15.3333 11.1144 15.781 10.6667 16.3333 10.6667C16.8856 10.6667
+      17.3333 11.1144 17.3333 11.6667Z" />
   {:else if name === "exclamation"}
     <path
       fill-rule="evenodd"
diff --git a/src/views/projects/Cob/Revision.svelte b/src/views/projects/Cob/Revision.svelte
index d779d855..bccf6650 100644
--- a/src/views/projects/Cob/Revision.svelte
+++ b/src/views/projects/Cob/Revision.svelte
@@ -11,12 +11,14 @@
   import Clipboard from "@app/components/Clipboard.svelte";
   import CommentComponent from "@app/components/Comment.svelte";
   import DiffStatBadge from "@app/components/DiffStatBadge.svelte";
+  import Dropdown from "@app/components/Dropdown.svelte";
   import ErrorMessage from "@app/components/ErrorMessage.svelte";
+  import Floating from "@app/components/Floating.svelte";
   import Icon from "@app/components/Icon.svelte";
   import InlineMarkdown from "@app/components/InlineMarkdown.svelte";
   import ProjectLink from "@app/components/ProjectLink.svelte";
-  import ThreadComponent from "@app/components/Thread.svelte";
   import SquareButton from "@app/components/SquareButton.svelte";
+  import ThreadComponent from "@app/components/Thread.svelte";
 
   export let authorId: string;
   export let baseUrl: BaseUrl;
@@ -141,14 +143,6 @@
         Revision {utils.formatObjectId(revisionId)}
       </span>
       <Clipboard text={revisionId} small />
-      {#if previousRevOid}
-        <ProjectLink
-          projectParams={{
-            search: `diff=${previousRevOid}..${revisionOid}`,
-          }}>
-          <SquareButton size="small">diff</SquareButton>
-        </ProjectLink>
-      {/if}
     </div>
     <div class="txt-small" />
     <div class="revision-data">
@@ -159,6 +153,31 @@
       <span class="layout-desktop txt-small">
         {utils.formatTimestamp(revisionTimestamp)}
       </span>
+      {#if previousRevOid}
+        <Floating>
+          <svelte:fragment slot="toggle">
+            <SquareButton size="small">
+              <Icon name="ellipsis-small" />
+            </SquareButton>
+          </svelte:fragment>
+          <svelte:fragment slot="modal">
+            <Dropdown
+              items={[
+                { title: previousRevOid, value: previousRevOid, badge: null },
+              ]}
+              selected={previousRevOid}>
+              <svelte:fragment slot="item" let:item>
+                <ProjectLink
+                  projectParams={{
+                    search: `diff=${item.value}..${revisionOid}`,
+                  }}>
+                  compare to {utils.formatObjectId(previousRevOid)}
+                </ProjectLink>
+              </svelte:fragment>
+            </Dropdown>
+          </svelte:fragment>
+        </Floating>
+      {/if}
     </div>
   </div>
   {#if expanded}

```

**Comment by xphoniex** [comment]
Yeah, I don't mind that, my only concern would be that it's hidden, "Compare" is pretty hidden in GitHub as well. 

I'd still prefer an icon next to clipboard icon, for now. 

Another thing, having it as an icon on the left side doesn't mess up the grid, but ellipsis on the right side does. (we simply can add an empty filler though)

Like both options, but slightly prefer the icon on left, let's see what others prefer.

**Comment by xphoniex** [comment]
Btw, github diff pages don't include either PR # in the url or show its description which is different than what we're doing here:
Screenshot from 2023-05-17 11-59-14

**Comment by rudolfs** [comment]
> Before the diff view, we are seeing the patch description, we can either keep that or show the Revision description there, or maybe the Revision and its previous one

I think we should show in the header clearly that we're comparing two revisions, so explicitly saying: Comparing PREVIOUS_SHA..THIS_SHA, just like in the GH screenshot you shared.

**Comment by rudolfs** [comment]
> Another thing, having it as an icon on the left side doesn't mess up the grid, but ellipsis on the right side does. (we simply can add an empty filler though)

We can show the context icon always and just disable it when there are no items in it. This way the grid would stay unchanged.

**Comment by xphoniex** [comment]
> > Before the diff view, we are seeing the patch description, we can either keep that or show the Revision description there, or maybe the Revision and its previous one
> 
> I think we should show in the header clearly that we're comparing two revisions, so explicitly saying: Comparing PREVIOUS_SHA..THIS_SHA, just like in the GH screenshot you shared.
> 
> Screenshot 2023-05-17 at 10 37 23

Looks good, can you add the patch diff too, if you have a working one?

**Comment by rudolfs** [comment]
> Looks good, can you add the patch diff too, if you have a working one?

I did that mockup by editing the dom in chrome : P

**Comment by cloudhead** [comment]
Nice! Some thoughts:

* Since the latest revision is open by default, I would also maybe have the button not hidden in a menu, but show up in the activity item where it says "cloudhead updated this patch to ..." -- I just think users won't generally look in the `...` menu initially.
* In the revision diff view, I would want to see the *revision ids* being diffed, not just the commits. Maybe this could be shown somewhere on the right of the tabs?

Otherwise looks great :+1:

**Comment by xphoniex** [comment]
> * In the revision diff view, I would want to see the _revision ids_ being diffed, not just the commits. Maybe this could be shown somewhere on the right of the tabs?

We can show that in-place of patch description, wdyt?

File: src/views/projects/Cob/Revision.svelte [hunk 1]
```diff
@@ -149,6 +154,36 @@
...
+          </svelte:fragment>
+          <svelte:fragment slot="modal">
+            <Dropdown
+              items={[
+                { title: previousRevOid, value: previousRevOid, badge: null },
+              ]}
+              selected={previousRevOid}>
+              <svelte:fragment slot="item" let:item>
+                <ProjectLink
+                  title="{item.value}..{revisionOid}"
```
Comment (xphoniex) [inline src/views/projects/Cob/Revision.svelte:172]: added this `title` cause originally it was only showing the first part of the hash.

**Comment by xphoniex** [comment]
> * Since the latest revision is open by default, I would also maybe have the button not hidden in a menu, but show up in the activity item where it says "cloudhead updated this patch to ..." -- I just think users won't generally look in the `...` menu initially.

I think this is more suited to a message like these after last revision card:

Screenshot from 2023-05-17 20-10-35

but we don't seem to have a system like that atm. maybe best to add in a separate PR?

File: src/views/projects/Patch.svelte [hunk 2]
```diff
@@ -308,6 +309,16 @@
...
+        {#if diff}
+          <ProjectLink
+            projectParams={{
+              search: `diff=${diff}`,
+            }}>
+            <SquareButton size="small" active={true}>
+              diff {diff.substr(0, 6)}..{diff.split("..")[1].substr(0, 6)}
+            </SquareButton>
+          </ProjectLink>
+        {/if}
```
[the remark this answers is not in the document]
Comment (xphoniex) [reply src/views/projects/Patch.svelte:321]: `diff 969ba8..423809` or `Diff 969ba8..423809`

[the remark this answers is not in the document]
Comment (rudolfs) [reply src/views/projects/Patch.svelte:321]: I'd say "Compare 969ba8..423809"

**Comment by xphoniex** [comment]
> > * Since the latest revision is open by default, I would also maybe have the button not hidden in a menu, but show up in the activity item where it says "cloudhead updated this patch to ..." -- I just think users won't generally look in the `...` menu initially.
> 
> I think this is more suited to a message like these after last revision card:
> 
> Screenshot from 2023-05-17 20-10-35
> 
> but we don't seem to have a system like that atm. maybe best to add in a separate PR?

any comments re this @rudolfs ? if you can add a timeline-like component like that connected to last card, I can add diff there as well.

File: src/views/projects/Patch.svelte (hunk 2 above)
[the remark this answers is not in the document]
Comment (xphoniex) [reply src/views/projects/Patch.svelte:321]: I didn't want to just copy GitHub's terminology, as for comparing commits, `diff`/`Diff` is something developers are familiar with and makes sense but kept revision message "compare to revision <hash>". wdyt?

[the remark this answers is not in the document]
Comment (cloudhead) [reply src/views/projects/Patch.svelte:321]: `Compare` or `Diff` works for me.

**Comment by cloudhead** [comment]
Perhaps we can keep it simple for this PR and just put the button somewhere easy to access, like the original screenshot, and then move it to the timeline?

**Comment by cloudhead** [comment]
> We can show that in-place of patch description, wdyt?

Not sure how that would look.. I think above the file makes the most sense, where it says `1 file changed...`. It could go to the other (right) side.

We probably shouldn't show the patch description there though, as it could be really long.

**Comment by xphoniex** [comment]
latest: 
Screenshot from 2023-05-24 16-58-12

File: src/views/projects/Cob/Revision.svelte [hunk 3]
```diff
@@ -142,13 +146,65 @@
...
+                  },
+                ]}>
+            <svelte:fragment slot="item" let:item>
+              <ProjectLink
+                title="{item.value}..{revisionOid}"
+                projectParams={{
+                  search: `diff=${item.value}..${revisionOid}`,
+                }}>
+                {#if item.value === projectHead}
+                  Compare to master ({utils.formatObjectId(projectHead)})
```
Comment (rudolfs) [inline src/views/projects/Cob/Revision.svelte:197]:
```diff
diff --git a/src/views/projects/Cob/Revision.svelte b/src/views/projects/Cob/Revision.svelte
index f7599228..e8374a19 100644
--- a/src/views/projects/Cob/Revision.svelte
+++ b/src/views/projects/Cob/Revision.svelte
@@ -24,6 +24,7 @@
   export let expanded: boolean = true;
   export let patchId: string;
   export let projectHead: string;
+  export let projectDefaultBranch: string;
   export let projectId: string;
   export let revisionBase: string;
   export let revisionId: string;
@@ -194,7 +195,9 @@
                   search: `diff=${item.value}..${revisionOid}`,
                 }}>
                 {#if item.value === projectHead}
-                  Compare to master ({utils.formatObjectId(projectHead)})
+                  Compare to {projectDefaultBranch} ({utils.formatObjectId(
+                    projectHead,
+                  )})
                 {:else if previousRevId}
                   Compare to previous revision ({utils.formatObjectId(
                     previousRevId,
diff --git a/src/views/projects/Patch.svelte b/src/views/projects/Patch.svelte
index 5943b0b6..159eab0f 100644
--- a/src/views/projects/Patch.svelte
+++ b/src/views/projects/Patch.svelte
@@ -67,6 +67,7 @@
   export let baseUrl: BaseUrl;
   export let patch: Patch;
   export let projectHead: string;
+  export let projectDefaultBranch: string;
   export let revision: string;
   export let currentTab: "activity" | "commits" | "files";
   export let diff: string | undefined = undefined;
@@ -372,11 +373,11 @@
         {#each timelineTuple as [revision, timelines], index}
           {@const previousRevision =
             index > 0 ? patch.revisions[index - 1] : undefined}
-
           <RevisionComponent
             {baseUrl}
             {projectId}
             {timelines}
+            {projectDefaultBranch}
             {projectHead}
             {...revision}
             on:reply={createReply}
diff --git a/src/views/projects/View.svelte b/src/views/projects/View.svelte
index 53c5959b..7269d701 100644
--- a/src/views/projects/View.svelte
+++ b/src/views/projects/View.svelte
@@ -272,6 +272,7 @@
             {patch}
             {baseUrl}
             projectId={project.id}
+            projectDefaultBranch={project.defaultBranch}
             projectHead={project.head}
             revision={activeRoute.params.view.params.revision ??
               latestRevision.id}

```

**Comment by rudolfs** [comment]
> In the revision diff view, I would want to see the revision ids being diffed, not just the commits. Maybe this could be shown somewhere on the right of the tabs?

I'd leave this for a follow-up, because it's not straightforward to pass in the revision ids to the sub-view.

**Review by rudolfs — APPROVED** [review]
:rocket:

**Pull request merged**